java/src/main/java/cucumber/runtime/java/JavaBackend.java

```diff
@@ -59,24 +60,25 @@ private Snippet createSnippet() {
      * @param resourceLoader
      */
     public JavaBackend(ResourceLoader resourceLoader) {
-        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        classFinder = new ResourceLoaderClassFinder(resourceLoader, classLoader);
-        methodScanner = new MethodScanner(classFinder);
-        objectFactory = ObjectFactoryLoader.loadObjectFactory(classFinder, Env.INSTANCE.get(ObjectFactory.class.getName()));
+        this(new ResourceLoaderClassFinder(resourceLoader, currentThread().getContextClassLoader()));
+    }
+
+    private JavaBackend(ClassFinder classFinder) {
+        this(loadObjectFactory(classFinder, Env.INSTANCE.get(ObjectFactory.class.getName())), classFinder);
     }
 
     public JavaBackend(ObjectFactory objectFactory) {
-        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        ResourceLoader resourceLoader = new MultiLoader(classLoader);
-        classFinder = new ResourceLoaderClassFinder(resourceLoader, classLoader);
-        methodScanner = new MethodScanner(classFinder);
-        this.objectFactory = objectFactory;
+        this(objectFactory,
+            new ResourceLoaderClassFinder(
+                new MultiLoader(
```

Comment:
You are right. I was missing a constructor `JavaBackend(ObjectFactory, ResourceLoader)`. This made look into the root cause, why do we have these constructors in the first place. It turns out they're only used by tests which needed them because they were poorly written. 

I've added a commit which improves these test somewhat.